### PR TITLE
Alias fix

### LINKED ISSUE #303: Aliases with bindings in middle of pipeline crash
Given the following alias:

```
Name: body
Visibility: user
Pipeline: echo $body
```

Running `echo foo | body` results in a stack trace in the Cog console with no error returned to the user in chat:

```
2016-02-25T22:59:08.0605  () [error] ** State machine #PID<0.2197.0> terminating
** Last message in was {:publish, "/bot/pipelines/b2ed1a6238564f97b7efefdd7a489f80/reply",
 "{\"signature\":\"34578976F146B3D44208C44BC263877E575AE8655BB2DAA866F1B86CE3DFFA5870FEBFF398E2DB495FF81FD72E8F7FA79E551E2B42C7EBD85F8061E788237A09\",\"id\":\"c1c6808f-8e66-4971-88f6-e6d17b8e031c\",\"data\":{\"template\":null,\"status_message\":null,\"status\":\"ok\",\"room\":null,\"bundle\":null,\"body\":{\"body\":[\"foo\"]}}}"}
** When State == :wait_for_command
**      Data  == %Cog.Command.Pipeline.Executor{context: nil,
 current: %Piper.Command.Ast.Invocation{args: [%Piper.Command.Ast.String{col: 6,
    line: 1, value: "foo"}], col: 1, command: "operable:echo", line: 1,
  options: [], redirs: []},
 current_bound: %Piper.Command.Ast.Invocation{args: ["foo"], col: 1,
  command: "operable:echo", line: 1, options: %{}, redirs: []},
 error_message: nil, error_type: nil, id: "b2ed1a6238564f97b7efefdd7a489f80",
 input: [], mq_conn: #PID<0.2198.0>, output: [],
 pipeline: %Piper.Command.Ast.Pipeline{abort_on_error: true, col: 10,
  finalized: nil,
  invocations: [%Piper.Command.Ast.Invocation{args: [%Piper.Command.Ast.String{col: 6,
      line: 1, value: "foo"}], col: 1, command: "operable:echo", line: 1,
    options: [], redirs: []},
   %Piper.Command.Ast.Invocation{args: [], col: 12, command: "user:body",
    line: 1, options: [], redirs: []}], line: 1, output_to_input: true,
  text: "|", type: :pipe}, redirects: [],
 remaining: [%Piper.Command.Ast.Invocation{args: [], col: 12,
   command: "user:body", line: 1, options: [], redirs: []}],
 request: %{"adapter" => "slack", "module" => "Elixir.Cog.Adapters.Slack",
   "reply" => "/bot/adapters/slack/send_message",
   "room" => %{"id" => "D0CMPQB63", "name" => "direct"},
   "sender" => %{"handle" => "imbriaco", "id" => "U03CU6VV5"},
   "text" => "echo foo | body"},
 scope: %Piper.Command.Bind.Scope{bindings: %{}, parent: nil, values: %{}},
 started: {1456, 459148, 569234},
 topic: "/bot/pipelines/b2ed1a6238564f97b7efefdd7a489f80",
 user: %Cog.Models.User{__meta__: #Ecto.Schema.Metadata<:loaded>,
  chat_handles: [%Cog.Models.ChatHandle{__meta__: #Ecto.Schema.Metadata<:loaded>,
    chat_provider: #Ecto.Association.NotLoaded<association :chat_provider is not loaded>,
    handle: "imbriaco", id: "3d46e80c-8bde-4ea0-9496-33e19a759b34",
    inserted_at: #Ecto.DateTime<2016-02-25T21:40:23Z>, provider_id: 1,
    updated_at: #Ecto.DateTime<2016-02-25T21:40:23Z>,
    user: #Ecto.Association.NotLoaded<association :user is not loaded>,
    user_id: "53df8c3d-51ee-4e60-bd2f-4e445ff30b5d"},
   %Cog.Models.ChatHandle{__meta__: #Ecto.Schema.Metadata<:loaded>,
    chat_provider: #Ecto.Association.NotLoaded<association :chat_provider is not loaded>,
    handle: "imbriaco", id: "efe3d3e8-cb2d-41dc-a183-50c3aebba5cb",
    inserted_at: #Ecto.DateTime<2016-02-26T02:35:32Z>, provider_id: 2,
    updated_at: #Ecto.DateTime<2016-02-26T02:35:32Z>,
    user: #Ecto.Association.NotLoaded<association :user is not loaded>,
    user_id: "53df8c3d-51ee-4e60-bd2f-4e445ff30b5d"}],
  direct_group_memberships: #Ecto.Association.NotLoaded<association :direct_group_memberships is not loaded>,
  email_address: "imbriaco@operable.io", first_name: "Mark",
  group_memberships: #Ecto.Association.NotLoaded<association :group_memberships is not loaded>,
  id: "53df8c3d-51ee-4e60-bd2f-4e445ff30b5d",
  inserted_at: #Ecto.DateTime<2016-02-25T21:40:21Z>, last_name: "Imbriaco",
  password: nil,
  password_digest: "$2b$14$LAWR5pP1FQ16kFxlgf8DYeTX0WDs2l6qmO2GZRLsRHg60N.Wo8r3.",
  permission_grants: #Ecto.Association.NotLoaded<association :permission_grants is not loaded>,
  permissions: #Ecto.Association.NotLoaded<association :permissions is not loaded>,
  role_grants: #Ecto.Association.NotLoaded<association :role_grants is not loaded>,
  roles: #Ecto.Association.NotLoaded<association :roles is not loaded>,
  tokens: #Ecto.Association.NotLoaded<association :tokens is not loaded>,
  updated_at: #Ecto.DateTime<2016-02-25T21:40:21Z>, username: "imbriaco"},
 user_permissions: ["operable:manage_commands", "operable:manage_groups",
  "operable:manage_permissions", "operable:manage_roles",
  "operable:manage_users"]}
** Reason for termination =
** {{:badmatch, {:not_found, "user:body"}},
 [{Cog.Command.Pipeline.Executor, :prepare_or_finish, 2,
   [file: 'lib/cog/command/pipeline/executor.ex', line: 542]},
  {:gen_fsm, :handle_msg, 7, [file: 'gen_fsm.erl', line: 518]},
  {:proc_lib, :init_p_do_apply, 3, [file: 'proc_lib.erl', line: 240]}]}

2016-02-25T22:59:08.0606  () [error] ** State machine #PID<0.2198.0> terminating
** Last message in was {:EXIT, #PID<0.2197.0>,
 {{:badmatch, {:not_found, "user:body"}},
  [{Cog.Command.Pipeline.Executor, :prepare_or_finish, 2,
    [file: 'lib/cog/command/pipeline/executor.ex', line: 542]},
   {:gen_fsm, :handle_msg, 7, [file: 'gen_fsm.erl', line: 518]},
   {:proc_lib, :init_p_do_apply, 3, [file: 'proc_lib.erl', line: 240]}]}}
** When State == :connected
**      Data  == {:state, #PID<0.2197.0>, '<0.2198.0>', {127, 0, 0, 1}, 1883, #Port<0.143406>,
 #PID<0.2199.0>,
 {:proto_state, #Port<0.143406>, '127.0.0.1:59208', 4, "MQTT",
  "emqttc_trogdor_6496478a901ac62f32e7", true, 90, false,
  {:mqtt_message, 0, false, false, :undefined, :undefined, :undefined},
  :undefined, :undefined, 2,
  %{"/bot/pipelines/b2ed1a6238564f97b7efefdd7a489f80/+" => 1}, %{}, %{}, %{},
  {:gen_logger, :lager_logger, 2}}, [{#PID<0.2197.0>, #Reference<0.0.1.18047>}],
 %{"/bot/pipelines/b2ed1a6238564f97b7efefdd7a489f80/+" => {1,
    [#PID<0.2197.0>]}}, [], [], %{}, 1, false,
 {:keepalive, #Port<0.143406>, :send_oct, 49, 90, {:keepalive, :timeout},
  #Reference<0.0.1.18045>}, 90, 60, 4, 8, :undefined, :tcp, :undefined,
 {:gen_logger, :lager_logger, 2}, [], []}
** Reason for termination =
** {{:badmatch, {:not_found, "user:body"}},
 [{Cog.Command.Pipeline.Executor, :prepare_or_finish, 2,
   [file: 'lib/cog/command/pipeline/executor.ex', line: 542]},
  {:gen_fsm, :handle_msg, 7, [file: 'gen_fsm.erl', line: 518]},
  {:proc_lib, :init_p_do_apply, 3, [file: 'proc_lib.erl', line: 240]}]}
```

### PR DESCRIPTION
**Pull request by mpeck**
This fixes crashes when aliases are in the middle of pipelines.

resolves #303

### CODE REVIEW

File: lib/cog/command/alias_expander.ex [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    case is_alias?(alias_name) do
+      true ->
+        expand(alias_name, [], limit, 0)
+      false ->
+        {:ok, :not_an_alias}
+    end
+  end
+
+  defp expand(alias_name, acc, limit, current_depth) do
+    cond do
```
Comment (vanstee) [inline lib/cog/command/alias_expander.ex]: You should be able to replace this `cond` with pattern matching in a guard: `when current_dept <= limit`

File: lib/cog/command/alias_expander.ex [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    end
+  end
+
+  defp get_alias("user:" <> name) do
+    case Repo.get_by(UserCommandAlias, name: name) do
+      nil ->
+        :not_found
+      user_alias ->
+        user_alias
+    end
```
Comment (vanstee) [inline lib/cog/command/alias_expander.ex]: Not sure if I like this or not, but you can write this a little more succinctly like this:

```elixir
with nil <- Repo.get_by(UserCommandAlias, name: name),
  do: :not_found
```

File: lib/cog/command/command_resolver.ex [hunk 3]
```diff
@@ -0,0 +1,50 @@
+defmodule Cog.Command.CommandResolver do
+
+  alias Cog.Repo
+  alias Cog.Queries.Command
+  alias Cog.Queries.Alias
```
Comment (vanstee) [inline lib/cog/command/command_resolver.ex]: This gets kinda confusing to me if you're calling `Alias` but it's really a Query module. I always think it's the model module when I see `Alias`.

File: lib/cog/command/command_resolver.ex [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    end
+  end
+  def find_bundle(name) do
+    SemanticError.new("#{inspect name}", :no_command)
+  end
+
+  defp get_alias_type(name) do
+    case Repo.all(Alias.user_alias_by_name(name)) do
+      [] ->
+        case Repo.all(Alias.site_alias_by_name(name)) do
```
Comment (vanstee) [inline lib/cog/command/command_resolver.ex]: This seems kinda expensive to just get the type. Wish there was an `Repo.exists?` function for stuff like this. I guess we could always write one. Might be nice for this function.

Comment (christophermaier) [inline lib/cog/command/command_resolver.ex]: Could also solve it with a single carefully-crafted query. Not sure off hand if it'd be directly doable via Ecto's query DSL or if we'd need either a straight SQL query or an SQL stored procedure.

File: lib/cog/queries/alias.ex [hunk 5]
```diff
@@ -30,4 +30,16 @@ defmodule Cog.Queries.Alias do
     select: a
   end
 
+  def user_alias_by_name(name) do
+    from a in UserCommandAlias,
+    where: a.name == ^name,
+    select: a
```
Comment (vanstee) [inline lib/cog/queries/alias.ex]: I think we're using a new enough version of Ecto to use this expression shorthand:

```
UserCommandAlias |> where(name: name)
```

File: lib/cog/command/pipeline/executor.ex [hunk 6]
```diff
@@ -532,14 +548,24 @@ defmodule Cog.Command.Pipeline.Executor do
...
+                {:next_state, :bind, %{state | current: h, remaining: t, input: ot, output: [], context: oh}, 0}
+            end
+          {:not_found, "user:" <> _user_alias} ->
             {:next_state, :bind, %{state | current: h, remaining: t, input: [], output: [], context: new_output}, 0}
-          "multiple" ->
-            [oh|ot] = new_output
-            {:next_state, :bind, %{state | current: h, remaining: t, input: ot, output: [], context: oh}, 0}
+          {:not_found, "site:" <> _site_alias} ->
+            {:next_state, :bind, %{state | current: h, remaining: t, input: [], output: [], context: new_output}, 0}
+          {:not_found, command} ->
```
Comment (vanstee) [inline lib/cog/command/pipeline/executor.ex]: We really need that refactor. I'm having a really hard time keeping this all in my head.

**Comment by kevsmith** [comment]
Looks good overall based on my local testing. Two changes I think we need to make and then this should be good to merge modulo feedback from others.

1. The error message `Whoops! An error occurred. name has already been taken` should read `Whoops! An error occurred. The alias name is already in use.`
1. We should prevent users from aliasing fully qualified names as the final alias name will not be what they expect:
screen shot 2016-02-26 at 2 15 05 pm

File: lib/cog/command/alias_expander.ex (hunk 2 above)
Comment (mpeck) [inline lib/cog/command/alias_expander.ex]: TIL: `with` That's quite useful.

File: lib/cog/command/command_resolver.ex (hunk 4 above)
Comment (mpeck) [inline lib/cog/command/command_resolver.ex]: It is expensive, but I couldn't think of another way to solve it. I figured we could make it work and then make it work better later.

Comment (vanstee) [inline lib/cog/command/command_resolver.ex]: I'm ok with that :+1: @mpeck Maybe just add a TODO to fix it or something.

File: lib/cog/command/alias_expander.ex (hunk 2 above)
Comment (mpeck) [inline lib/cog/command/alias_expander.ex]: The with thing is cool, but I think in this case it's actually clearer to just use a case.

Comment (vanstee) [inline lib/cog/command/alias_expander.ex]: Yeah I was undecided about it, that's totally cool.

File: lib/cog/command/command_resolver.ex (hunk 3 above)
Comment (mpeck) [inline lib/cog/command/command_resolver.ex]: @vanstee how about `alias Cog.Queries.Alias, as: AliasQuery`?

Comment (vanstee) [inline lib/cog/command/command_resolver.ex]: @mpeck I've been doing `alias Cog.Queries` and then calling `Queries.Alias`, but that would also work.

File: lib/cog/queries/alias.ex (hunk 5 above)
Comment (mpeck) [inline lib/cog/queries/alias.ex]: @vanstee what do I need to import to get `with/2`?

Comment (vanstee) [inline lib/cog/queries/alias.ex]: It's a macro defined in `Ecto.Query`.

File: lib/cog/command/alias_expander.ex (hunk 1 above)
Comment (mpeck) [inline lib/cog/command/alias_expander.ex]: Good call.

**Comment by mpeck** [comment]
@kevsmith preventing aliases from being created with a fully qualified name is an alias management issue. Should that warrant another PR?

File: lib/cog/queries/alias.ex (hunk 5 above)
Comment (mpeck) [inline lib/cog/queries/alias.ex]: oh duh. I was reading `where` and typing `with` :|

**Comment by mpeck** [comment]
Actually now that I think about it. I think both of your issues are on the alias management side.

**Comment by kevsmith** [comment]
That's fine. Open an issue so we don't forget about it and assign it to 0.3.0.

**Pull request merged**